rfcs/v3/rest_json_api.md

```diff
@@ -0,0 +1,258 @@
+# REST API Output
+
+Hasura v3 will auto-generate a [JSON:API](https://jsonapi.org/) spec-compliant REST API output.
+Hasura v3 will only generate these for DDN models. Commands are out of scope of this RFC.
+
+## Fetching Data
+
+Data, including resources and relationships, can be fetched by sending a GET request to an endpoint.
+
+Responses can be further refined with the optional features described below.
+
+### Fetching Resources
+
+A server MUST support fetching resource data for every URL provided as:
+
+* a self link as part of the top-level links object
+* a self link as part of a resource-level links object
+* a related link as part of a relationship-level links object
+
+### Selecting
+
+JSON:API supports selecting specific fields of a resource and specific fields of its related
+resources.
+
+An endpoint **MAY** also support an `include` query parameter to allow the client to customize which
+related resources should be returned.
+
+A client **MAY** request that an endpoint return only specific
+[fields](https://jsonapi.org/format/#document-resource-object-fields) in the response on a per-type
+basis by including a `fields[TYPE]` query parameter. They are called [sparse
+fieldsets](https://jsonapi.org/format/#fetching-sparse-fieldsets).
+
+```
+GET /articles/1?include=comments
+
+GET /articles/1?include=comments.author
+
+GET /articles/1?include=comments.author,ratings
+
+GET /articles/1/relationships/comments?include=comments.author
+
+GET /articles?include=author&fields[articles]=title,body&fields[people]=name // sparse fieldsets
+```
+
+### Select Many
+
+```
+GET /<modelname>
+
+GET /articles
+```
+
+#### Filtering
+
+JSON:API is agnostic about the syntax and implementation of the server. The only requirement is to
+have a `filter` query parameter.
+
+So we propose the following for complex filtering use-cases.
+
+```
+GET /<modelname>?filter[field1]=val1&filter[field2]=val2,val3&filter[field3.reltarget_field1]=val4
+
+GET /comments?filter[post]=1,2&filter[author.id]=12
+
+GET /movies?filter={"$or":[{"name":{"$eq":"Braveheart"}},{"name":{"$eq":"Gladiator"}}]} // url encoded
+```
+
+Both the above syntax are equivalent. If you notice, the last syntax is equivalent to what Hasura
+generates for its GraphQL API.
+
+Hasura will also generate an OpenAPI schema for complex filtering on nested types. As a result, any
+auto-generated client would get full auto-complete and type-checking on any nested filtering
+clauses. Check out <https://github.com/hasura/json-api-experiments> to see this in action.
```

Comment:
I must have missed that we're able to get this working in a well-typed fashion. Nice!